File: CS214-Programming Language Concepts/projects/10/c++/birds.cpp
/* birds.cpp illustrates inheritance and polymorphism.
 *
 * Begun by: Dr. Adams, CS 214 at Calvin College.
 * Completed by: Joshua Maguire
 * Date: 4/29/14
 */

#include <iostream>       // cin, cout, <<, >>
using namespace std;

#include "Bird.h"         // Bird
#include "Owl.h"          // Owl
#include "Duck.h"         // Duck
#include "Goose.h"        // Goose

int main() {
  Bird * birdPtr0 = NULL,
       * birdPtr1 = NULL,
       * birdPtr2 = NULL,
       * birdPtr3 = NULL;

  birdPtr0 = new Bird("Hawkeye");
  birdPtr0->print(); cout << endl;
  
  birdPtr1 = new Duck("Donald");
  birdPtr1->print(); cout << endl;

  birdPtr2 = new Goose("Mother Goose");
  birdPtr2->print(); cout << endl;

  birdPtr3 = new Owl("Woodsey");
  birdPtr3->print(); cout << endl;

  delete birdPtr0;
  delete birdPtr1;
  delete birdPtr2;
  delete birdPtr3;
}

   inline Bird::Bird(const string & name) {
     myName = name;
   }
   
   inline string Bird::name() const {
     return myName;
   }
   
   inline string Bird::call() const {
     return "Squaaaaaaawk!";
   }
   
   inline string Bird::className() const {
     return "Bird";
   }
   
   inline void Bird::print(ostream & out) const {
      out << name() 
          << ' ' 
          << className() 
          << " says " 
          << call();
   } 
   
   
   

File: CS214-Programming Language Concepts/projects/10/project10/ruby/birds.rb
# birds.rb | Tests out the menagerie of bird classes.
#
# Begun by: Dr. Adams, for CS 214 at Calvin College.
# Completed by:Joshua Maguire
# Date: 5/4/14
######################################################

#in this version of ruby, require tries to find files in the load path.
#thats why I use require_relative. Also, ruby is smart enough to now need .rb
#This change is also in Owl, Duck, Duck, and Goose. 
require_relative 'Duck' 
require_relative 'Goose'
require_relative 'Owl'

bird0 = Bird.new "Hawkeye"
bird0.print

bird1 = Duck.new "Donald"
bird1.print

bird2 = Goose.new "Mother"
bird2.print

bird3 = Owl.new "Woodsey"
bird3.print